bf2fc6cc711aee1a0c2a/kas-fleet-manager
# feat: add markdown link checking

## Description

This PR adds the ability for developers to check for broken link references in markdown files locally and it also integrates markdown link checking into the CI/CD process of this repository.

### Details

#### Local development

This set of changes adds two Makefile targets
* markdown-link-check-install: installs the markdown-link-check
  tool which is able to check whether links are alive
* lint/markdown-link-check: Runs markdown-link-check checking
  all .md (markdown) files in the project. The bin directory is
  excluded

markdown-link-check is able to check:
* URLs
* Relative links to other files in the projects
* `mailto:` links

Additionally, a markdown-link-check configuration file
`.markdown-link-check/markdown-link-check-config.json` relative
to the project's root has been added. This file can be used
when running the markdown-link-check tool with the `-c` flag
referencing it. The `lint/markdown-link-check` Makefile target
makes use of it automatically.

Some URL patterns have been configured to be ignored in the
markdown-link-check configuration file. This URL patterns
correspond to URLs that resolve to private IP addresses or that
are private GitHub/GitLab repositories.

#### CI/CD process

Two GitHub workflows have been added that perform Markdown links
checking:
* .github/workflows/markdown-link-check-pr.yaml: Executed
  in each PR.
* .github/workflos/markdown-link-check-weekly.yaml: Executed
  Weekly on Mondays at 07:00 UTC.

All the specified md files and directories in the workflows are
intentionally checked, independently on whether they have been
modified or not. The reason for this is that even if a given .md
file has not been modified, its links can become broken. For example,
if a PR moves .md files from directories and some other .md files
contain relative links referencing the old path of the moved files,
then the links become broken even though the file containing the
references not being modified.

## Verification Steps
* The CI task to check markdown links is executed and passes
* Run markdown link checking locally by running `make lint/markdown-lint-check`. No errors should be reported.


## Checklist (Definition of Done)
